swift-corelibs: set NIX_COREFOUNDATION_RPATH in a hook

reckenrode opened the pull request:
###### Description of changes

Closes #230870. Thanks to @eliasnaur for the test case and for rasining awareness and to @veprbl for the work done on #111385.

This takes a slightly different approach from those two PRs. The hook is set unconditionally. The stdenv bootstrap doesn’t really need CF at all, so setting the hook is harmless. This simplifies things.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

reckenrode committed:
swift-corelibs: set NIX_COREFOUNDATION_RPATH in a hook

Closes #230870. Thanks to @eliasnaur for the test case and for rasining
awareness and to @veprbl for the work done on #111385.

This takes a slightly different approach from those two PRs. The hook is
set unconditionally. The stdenv bootstrap doesn’t really need CF at all,
so setting the hook is harmless. This simplifies things.

veprbl commented:
> Closes https://github.com/NixOS/nixpkgs/pull/230870

I don't see how this can fix darwin-to-linux cross.

reckenrode commented:
> > Closes #230870
> 
> I don't see how this can fix darwin-to-linux cross.

The other part is in the stdenv rework that drops setting `NIX_COREFOUNDATION_RPATH` from the `preHook`. I can do that here even though it’ll just be replaced by the rework, so this PR is self-contained and complete.

reckenrode committed:
darwin.stdenv: drop NIX_COREFOUNDATION_RPATH from preHook

Rely on swift-corefoundation to set this in its hook. That way, it will
be dropped from the `extraBuildInputs` during a cross build.

reckenrode commented:
@veprbl Sorry about that. I pushed a commit that makes this self-contained. This PR should address the cross issue properly now independently of the rework PR (which is blocked by #234859).

toonn commented on pkgs/os-specific/darwin/swift-corelibs/pure-corefoundation-hook.sh:
```diff
@@ -0,0 +1,7 @@
+usePureCoreFoundation() {
+# Avoid overriding value set by the impure CF
+    if [ -z "${NIX_COREFOUNDATION_RPATH:-}" ]; then
+        export NIX_COREFOUNDATION_RPATH=@out@/Library/Frameworks
```
I'm not very familiar with setup hooks. `@out@` is substituted right before running them?

veprbl replied:
It's substituted at build time.

toonn reviewed (commented):
Just one question for clarification. Given the rebuilds staging is appropriate I'll try doing a build anyway though but that'll have to be tomorrow.

veprbl commented:
Result of `nixpkgs-review pr 234861` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hello</li>
  </ul>
</details>

veprbl commented:
Result of `nixpkgs-review pr 234861` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hello</li>
  </ul>
</details>

toonn commented:
I'm still testing this. I don't expect issues so feel free to merge if you're confident in the changes.
Just wanted to ping to say I'm not neglecting this so the PR isn't "stuck."

toonn reviewed (approved):
Was finally able to reproduce on the staging commit, not sure what I was doing wrong before.

LGTM

eliasnaur commented:
Thank you! I can confirm that this was the final piece of the puzzle in my quest towards bit-for-bit reproducibility: my project now builds the exact same bits across all 4 combinations of Linux/macOS and x86_64/M1.
